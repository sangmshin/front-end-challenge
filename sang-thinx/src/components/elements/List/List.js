import React, { useRef, useState, useEffect } from 'react';
import { array, func, string } from 'prop-types';
import { ListItem } from 'components/elements';
import styles from './List.module.scss';
import cx from 'classnames';

export const List = ({
  items,
  onSelectItem,
  selectedItem,
}) => {
  const [shouldListOpenUpward, openUpward] = useState(false);

  const listElement = useRef();
  const listItems = items.map(item => (
    <ListItem
      item={item}
      key={item}
      onSelect={onSelectItem}
      selectedItem={selectedItem}
    />
  ));

  useEffect(() => {
    const elementPositionOnScreen = listElement.current.getBoundingClientRect().top;
    const enoughSpaceForDropdown = window.innerHeight - listElement.current.clientHeight;
    const isEnoughSpaceForDropdown = elementPositionOnScreen < enoughSpaceForDropdown;

    if (!isEnoughSpaceForDropdown) openUpward(true);
    else openUpward(false);
  }, []);

  const listClass = cx(styles.root, {
    [styles.upward]: shouldListOpenUpward
  });
  
  return (
    <ul
      ref={listElement}
      className={listClass}
      role='listbox'
    >
      {listItems}
    </ul>
  );
};

List.propTypes = {
  items: array.isRequired,
  onSelectItem: func.isRequired,
  selectedItem: string,
}

List.defaultProps = {
  selectedItem: null,
};

export default List;