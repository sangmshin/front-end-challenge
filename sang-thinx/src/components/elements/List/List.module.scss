@import 'variables';

.root {
  position: absolute;
  top: -1px;
  left: -1px;
  width: 100%;
  background-color: $primary-lightgrey;
  border: 1px solid black;

  &.upward{
    top: unset;
    bottom: -1px;
  }
}