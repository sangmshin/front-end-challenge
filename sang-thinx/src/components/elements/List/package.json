{
  "name": "List",
  "main": "./List.js"
}
