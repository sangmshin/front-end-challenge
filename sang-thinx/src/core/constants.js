export const COLOR_HEXCODE_MAP = {
  beige: '#f2d9a5',
  black: '#000',
}