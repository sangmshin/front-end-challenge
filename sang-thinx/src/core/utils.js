const SMALL = 768; /* Small screen / tablet */
const MEDIUM = 1024; /* Medium screen / desktop */
const LARGE = 1200; /* Large screen / wide desktop */
let BREAKPOINT;

/**
 * Finds the current matched media query breakpoint
 *
 * @returns {String} matched media query breakpoint [large, medium, small, xsmall]
 */
export function getCurrentBreakPoint() {
  if (!BREAKPOINT) {
    BREAKPOINT = {
      large: window.matchMedia(`(min-width: ${LARGE}px)`),
      medium: window.matchMedia(`(min-width: ${MEDIUM}px)`),
      small: window.matchMedia(`(min-width: ${SMALL}px)`),
      xsmall: window.matchMedia(`(max-width: ${SMALL - 1}px)`)
    };
  }

  return Object.keys(BREAKPOINT).find(breakpoint => BREAKPOINT[breakpoint].matches);
}

/**
 * Determines if the data is array and has at least one or more items
 * @return {boolean}
 */
export const isArrayWithLength = arr => Array.isArray(arr) && arr.length > 0;

/**
 * A function with no operation. (for testing)
 */
export const noop = () => null;
