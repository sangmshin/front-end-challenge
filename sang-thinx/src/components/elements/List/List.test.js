import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { List } from './List';
import { noop } from 'core/utils';

const props = {
  items: ['New York', 'Canada'],
  onSelectItem: noop,
  selectedItem: 'New York'
}

const component = shallow(<List {...props} />);

describe('Component :: List', () => {
  describe('#render', () => {
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('#number of items', () => {
    it('should match the number of items with "ListItem" component', () => {
      const numberOfItems = props.items.length;
      expect(component.find('ListItem').length).toBe(numberOfItems);
    });
  });
});
