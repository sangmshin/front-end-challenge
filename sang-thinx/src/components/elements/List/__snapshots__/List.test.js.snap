// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: List #render should match snapshot 1`] = `
<ul
  className="root"
  role="listbox"
>
  <ListItem
    item="New York"
    key="New York"
    onSelect={[Function]}
    selectedItem="New York"
  />
  <ListItem
    item="Canada"
    key="Canada"
    onSelect={[Function]}
    selectedItem="New York"
  />
</ul>
`;
